### ISSUE TITLE
SC2015: `A && B || true` should be considered OK

### ISSUE THREAD

**Issue by ShamrockLee** [koalaman/shellcheck#2789]
#### For bugs
- Rule Id: SC2015
- My shellcheck version (`shellcheck --version` or "online"): 0.9.0
- [X] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/usr/bin/env bash

FOO=1
[ "$FOO" -eq 1 ] && echo "yes" || true
```

#### Here's what shellcheck currently says:

```
[Line 4:](javascript:setPosition(4, 18))
[ "$FOO" -eq 1 ] && echo "yes" || true
                 ^-- [SC2015](https://www.shellcheck.net/wiki/SC2015) (info): Note that A && B || C is not if-then-else. C may run when A is true.

```

#### Here's what I wanted or expected to see:

ShellCheck should not throw this warning when C is `true`. The developer just want the return value to be 0, and it doesn't matter if `true` is executed or not.

**Comment by TinCanTech** [comment]
[ "$FOO" -eq 1 ] && echo "yes"
    :

**Comment by ShamrockLee** [comment]
> [ "$FOO" -eq 1 ] && echo "yes"
> :

With `set -e`, the script will fail before no-op.

**Comment by TinCanTech** [comment]
`set -e` is not a condition of your original post.

Also:
```
tct@home:~$ cat f
#!/bin/bash

set -e

[ $foo ] && echo ok
:
tct@home:~$ ./f
tct@home:~D$ echo $?
0
tct@home:~$ foo=1 ./f
ok
tct@home:~D$ echo $?
0
```
Also, `set -e` is the work of the Devil.

**Comment by wileyhy** [comment]
It is good practice to become adept and comfortable with writing error handlers

**Comment by TinCanTech** [comment]
It isn't that `shellcheck` is wrong:
```
tct@home:~$ unset -v A B; A=1; B=1; [ $A ] && [ $B ] || echo C
tct@home:~$ unset -v A B; A=1; [ $A ] && [ $B ] || echo C
C
tct@home:~$ unset -v A B; B=1; [ $A ] && [ $B ] || echo C
C
```
Quote SC2015: ... _C **may** run when A is true_.

**Comment by wileyhy** [comment]
#!/bin/bash
set -e
a=1
b=2
{ [ -n "$a" ] && [ -n "$b" ]; } 
( a="$1"; b="$2";[ -z "$a" ] || [ -z "$b" ]; )

**Comment by ShamrockLee** [comment]
> Also:\n\ntct@home:~$ cat f\n#!/bin/bash\n\nset -e\n\n[ $foo ] && echo ok\n:\ntct@home:~$ ./f\ntct@home:~D$ echo $?\n0\ntct@home:~$ foo=1 ./f\nok\ntct@home:~D$ echo $?\n0

I have never thought about that. Thank you for pointing it out.

**Comment by papb** [comment]
`set -e` is not applied to an expression of the form `A && B`, so `A && B || true` is the same as `A && B`.
